src/algorithms/local_search/swap_star_utils.h

```diff
@@ -0,0 +1,353 @@
+                                   s_rank);
+
+      auto swap_choice_options =
+        std::set<SwapChoice, decltype(SwapChoiceCmp)>(SwapChoiceCmp);
+
+      // Options for in-place insertion in source route include
+      // in-place insertion in target route and other relevant
+      // positions from target_insertions.
+      const Gain in_place_source_insertion_gain =
+        target_remove_gain - source_in_place_delta;
+      const Gain in_place_target_insertion_gain =
+        source_remove_gain - target_in_place_delta;
+
+      Gain current_gain =
+        in_place_target_insertion_gain + in_place_source_insertion_gain;
+      if (current_gain > best_gain) {
+        swap_choice_options.insert(
+          {current_gain, s_rank, t_rank, s_rank, t_rank});
+      }
+
+      for (const auto& ti : target_insertions) {
+        if ((ti.rank != t_rank) and (ti.rank != t_rank + 1) and
+            (ti.cost != std::numeric_limits<Gain>::max())) {
+          const Gain t_gain = source_remove_gain - ti.cost;
+          current_gain = in_place_source_insertion_gain + t_gain;
+          if (current_gain > best_gain) {
+            swap_choice_options.insert(
+              {current_gain, s_rank, t_rank, s_rank, ti.rank});
+          }
+        }
+      }
+
+      // Options for other relevant positions for insertion in source
+      // route (from source_insertions) include in-place insertion in
+      // target route and other relevant positions from
+      // target_insertions.
+      for (const auto& si : source_insertions) {
+        if ((si.rank != s_rank) and (si.rank != s_rank + 1) and
+            (si.cost != std::numeric_limits<Gain>::max())) {
+          const Gain s_gain = target_remove_gain - si.cost;
+
+          current_gain = s_gain + in_place_target_insertion_gain;
+          if (current_gain > best_gain) {
+            swap_choice_options.insert(
+              {current_gain, s_rank, t_rank, si.rank, t_rank});
+          }
+
+          for (const auto& ti : target_insertions) {
+            if ((ti.rank != t_rank) and (ti.rank != t_rank + 1) and
+                (ti.cost != std::numeric_limits<Gain>::max())) {
+              const Gain t_gain = source_remove_gain - ti.cost;
+              current_gain = s_gain + t_gain;
+              if (current_gain > best_gain) {
+                swap_choice_options.insert(
+                  {current_gain, s_rank, t_rank, si.rank, ti.rank});
+              }
+            }
+          }
+        }
+      }
+
+      assert(swap_choice_options.size() <= 16);
+
+      for (const auto& sc : swap_choice_options) {
+        // Browse interesting options by decreasing gain and check for
+        // validity.
+        const auto s_insert = get_insert_range(source.route,
+                                               s_rank,
+                                               target.route[t_rank],
+                                               sc.insertion_in_source);
```

Comment:
I think this is the actual pain point for computing times since here `get_insert_range` populates a new `std::vector` that serve as insertion range for the validity checks down the line.

The reason this is required in the first place is that the actual range is mostly a part of the source route, but with the target node added either in front or back depending on respective insertion ranks.

I don't see a way to solve this better for now, but I hope there is. Maybe at some point using ranges to be able to iterate without actually building the range?